Add build vit models notes (#256)
Oneflow-Inc/libai: docs/source/notes/How_to_build_vision_transformer_model_in_LiBai.md

> # Detailed instruction on building vision transformer model in LiBai

Author: 最终修改的标题~

> # Detailed instruction on how to build vision transformer model from scratch in LiBai

Author: 在 @khloe-zhang 的建议下修改了一下标题~，和内容更贴切

Reviewer: 这个标题会不会有点太长了

Author: 好像确实，我改成`Detailed instruction on vision transformer model in LiBai`吧

> It's easy for the users to use libai's built-in layers, e.g., [libai.layers](https://libai.readthedocs.io/en/latest/modules/libai.layers.html), to build the `transformer-based` models. Let's take a deep look into the process of building vision transformer model in LiBai.

Reviewer 2: ```suggestion
It's easy for the users to build the `transformer-based` models by using the libai's built-in [layers](https://libai.readthedocs.io/en/latest/modules/libai.layers.html). Let's take a deep look into the process of building vision transformer model in LiBai.
```

Author: 已改~

## Model Architecture
**Vision Transformer** was released with paper [An Image is Worth 16x16 Words: Transformers for Image Recognition at Scale](https://arxiv.org/abs/2010.11929) by Alexey Dosovitskiy, Lucas Beyer, Alexander Kolesnikov, Dirk Weissenborn, Xiaohua Zhai, Thomas Unterthiner, Mostafa Dehghani, Matthias Minderer, Georg Heigold, Sylvain Gelly, Jakob Uszkoreit, Neil Houlsby.

**Vision Transformer** model contains three parts: `Patch Embedding` + `Transformer Block` + `Linear Classification Head`, which can be summarized as the following picture:

![](./assets/vision_transformer.png)

## A simple torch implementation of Vision Transformer

> Let's take a look at the pytorch implementation of ViT models, this code is modified from [timm.models.vision_transformer](https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/vision_transformer.py):

Reviewer 2: ```suggestion
The following code shows the pytorch implementation of ViT models which is modified from [timm.models.vision_transformer](https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/vision_transformer.py):
```

Author: 已改~

        # weight init
        trunc_normal_(self.pos_embed, std=0.02)
        trunc_normal_(self.cls_token, std=0.02)
        self.apply(self._init_weights)

    def _init_weights(self, m):
        if isinstance(m, nn.Linear):
            trunc_normal_(m.weight, std=0.02)
            if m.bias is not None:
                nn.init.constant_(m.bias, 0)
        elif isinstance(m, nn.LayerNorm):
            nn.init.constant_(m.bias, 0)
            nn.init.constant_(m.weight, 1.0)

    def forward_features(self, x):
        # patch embedding
        x = self.patch_embed(x)

        cls_token = self.cls_token.expand(x.shape[0], -1, -1)
        x = torch.cat((cls_token, x), dim=1)

        # position embedding
        pos_embed = self.pos_embed.expand(x.shape[0], -1, -1)
        x = self.pos_drop(x + pos_embed)

        # transformer block
        x = self.blocks(x)
        return x

    def forward_head(self, x):
        # only use cls token for classification
        x = self.norm(x)
        outcome = x[:, 0]
        return self.head(outcome)
    
    def forward(self, x):
        x = self.forward_features(x)
        x = self.forward_head(x)
        return x
```

> We further decoupled the forward function into `forward_features` and `forward_head`:

Reviewer 2: ```suggestion
We have further decoupled the forward function into `forward_features` and `forward_head`:
```

Author: 已改~

- `forward_features`: extract the image features using `patch_embed` layer and a stack of `transformer` block
- `forward_head`: take the `cls_token` of each sample and use `nn.Linear` for classification

## Implement 3D parallel Vision Transformer in LiBai
In this section, we will show users how to use [libai.layers](https://libai.readthedocs.io/en/latest/modules/libai.layers.html) to build 3D parallel Vision Transformer model with only 100+ lines of code, which is modified from [libai.models.vision_transformer](https://github.com/Oneflow-Inc/libai/blob/main/libai/models/vision_transformer.py)

> Here is the LiBai implementation of Vision Transformer models, users only need to replace the pytorch modules with libai.layers as follows:

Reviewer 2: ```suggestion
Here is the LiBai implementation of Vision Transformer models, users only need to replace the pytorch modules with the corresponding libai.layers as follows:
```

Author: 已改~

                module_block.config.stage_id = dist_utils.get_layer_stage_id(module_block.layer_idx)

        # Set pos_embed and cls_token stage id
        model.pos_embed.config.stage_id = dist_utils.get_layer_stage_id(0)
        model.cls_token.config.stage_id = dist_utils.get_layer_stage_id(0)
        model.pos_drop.config.stage_id = dist_utils.get_layer_stage_id(0)
        model.norm.config.stage_id = dist_utils.get_layer_stage_id(-1)
        model.head.config.stage_id = dist_utils.get_layer_stage_id(-1)
        model.loss_func.config.stage_id = dist_utils.get_layer_stage_id(-1)
```

## Details about LiBai's implementation of Vision Transformer Model

**1. Replace nn.Module with libai.layers**

LiBai has already implemented `PatchEmbedding`, `TransformerLayer`, `Linear`, `LayerNorm` layers, users only need to replace the module in torch vision transformer models to convert a torch model into libai's style:
  - `Block` -> `libai.layers.TransformerLayer`
  - `nn.Linear` -> `libai.layers.Linear`
  - `nn.LayerNorm` -> `libai.layers.LayerNorm`
  - `PatchEmbed` -> `libai.layers.PatchEmbedding`

**2. Manually set the sbp signature of `cls_token` and `pos_embed`**

> In order to fit different parallel mode in LiBai, users need to manually set the **sbp signature** of the layers built with `nn.Parameter` which are not implemented in LiBai, like `cls_token` and `pos_embed` in Vision Transformer:

Reviewer 2: ```suggestion
In order to fit different parallel mode in LiBai, users must manually set the **sbp signature** for all the parameters and buffers of those layers not implemented in LiBai, like `cls_token` and `pos_embed` in Vision Transformer:
```

Author: 已改~

> In order to fit different parallel mode in LiBai, users must manually set the **sbp signature** for all the parameters and buffers of those layers not implemented in LiBai, like `cls_token` and `pos_embed` in Vision Transformer:

Reviewer 3: 这里要不要给个SBP的说明文档链接

Author: 已经修改~

```python
import oneflow as flow
import oneflow.nn as nn
import libai.utils.distributed as dist

self.cls_token = nn.Parameter(
    flow.zeros(
        1, 1, embed_dim,
        sbp=dist.get_nd_sbp([flow.sbp.broadcast, flow.sbp.broadcast]),
        placement=dist.get_layer_placement(0),
    )
)
self.pos_embed = nn.Parameter(
    flow.zeros(
        1, num_patches+1, embed_dim,
        sbp=dist.get_nd_sbp([flow.sbp.broadcast, flow.sbp.broadcast]),
        placement=dist.get_layer_placement(0),)
)
```

> - use `dist.get_nd_sbp([flow.sbp.broadcast, flow.sbp.broadcast])` to make the `cls_token` and `pos_embed` broadcast in each GPU groups.

Author: 已改~

**3. Use `to_global()` function to update the sbp signature of `cls_token` and `pos_embed` during forward function**

In forward function, `cls_token` and `pos_embed` will be expanded to fit the input size, for efficiency, we can use `to_global()` function to match the `cls_token` and `pos_embed` sbp signature with the input sbp signature like this:
```python
def forward_features(self, x):
    cls_token = self.cls_token.expand(
        x.shape[0], -1, -1
    )
    # use to_global to update the sbp signature of cls_token
    cls_token = cls_token.to_global(sbp=x.sbp, placement=cls_token.placement)
    x = flow.cat((cls_token, x), dim=1)

    # use to_global to update the sbp signature of pos_embed
    pos_embed = self.pos_embed.expand(x.shape[0], -1, -1)
    pos_embed = pos_embed.to_global(sbp=x.sbp, placement=pos_embed.placement)
```

> **4. Manually set the stage id for pipeline parallel training**

Author: 这个部分的表述再帮忙看看 @L1aoXingyu @CPFLAME @Ldpe2G

> Most of the built-in layers in LiBai has the arg named `layer_idx` for pipeline parallel settings, to configure a 1F1B pipeline parallel model, users should manually set the stage id for each layers in the models, with the help of `layer_idx` we can simply get a pipeline parallel Vision Transformer models like:

Reviewer 3: 这里感觉可以再加一句解释一下,  大意应该是  stage id 是为了分配当前的layer到 对应的卡上, 最终形成流水.

Author: 已经修改，可以帮忙看看表述是否有问题~
